src/python/alog/protocols.py

```diff
@@ -0,0 +1,576 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+################################################################################
+"""Protocols to assist IDEs with code completion for the additional methods which
+Alchemy Logger overrides on the logging.Logger instances.
+"""
+
+import sys
+from logging import Filter, LogRecord
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    ClassVar,
+    List,
+    Mapping,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+from .alog import _Level
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Protocol
+else:
+    from typing import Protocol
+
+if TYPE_CHECKING:
+    from logging import (
+        Handler,
+        Logger,
+        Manager,
+        RootLogger,
+        _ArgsType,
+        _ExcInfoType,
+        _SysExcInfoType,
+    )
+
+    from typing_extensions import Self
+
+
+class _SupportsFilter(Protocol):
+    def filter(self, __record: LogRecord) -> bool: ...
+
+
+_FilterType = Union[Filter, Callable[[LogRecord], bool], _SupportsFilter]
+
+
+class FilterProtocol(Protocol):
+    """
+    A base class for loggers and handlers which allows them to share
+    common code.
+    """
+
+    filters: List[_FilterType]
+
+    def __init__(self) -> None:
+        """
+        Initialize the list of filters to be an empty list.
+        """
+
+    def addFilter(self, filter: _FilterType) -> None:  # pylint: disable=invalid-name
+        """
+        Add the specified filter to this handler.
+        """
+
+    def removeFilter(self, filter: _FilterType) -> None:  # pylint: disable=invalid-name
+        """
+        Remove the specified filter from this handler.
+        """
+
+    def filter(self, record: LogRecord) -> bool:  # type: ignore
+        """
+        Determine if a record is loggable by consulting all the filters.
+
+        The default is to allow the record to be logged; any filter can veto
+        this and the record is then dropped. Returns a zero value if a record
+        is to be dropped, else non-zero.
+
+        .. versionchanged:: 3.2
+
+           Allow filters to be just callables.
+        """
+
+    if sys.version_info >= (3, 12):
+
+        def filter(self, record: LogRecord) -> Union[bool, LogRecord]:  # type: ignore pylint: disable=function-redefined
```

Comment:
Could you put the above def behind an `else` and avoid the `disable=function-redefined`?